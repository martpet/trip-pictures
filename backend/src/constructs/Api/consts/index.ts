export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
