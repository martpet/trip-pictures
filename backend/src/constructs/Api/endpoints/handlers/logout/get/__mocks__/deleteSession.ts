export const deleteSession = vi
  .fn()
  .mockName('deleteSession')
  .mockResolvedValue('dummyRefreshToken');
