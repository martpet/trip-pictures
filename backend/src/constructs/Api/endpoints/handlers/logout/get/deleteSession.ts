import {
  DeleteCommand,
  DynamoDBClient,
  DynamoDBDocumentClient,
  SessionsTableItem,
  sessionsTableOptions,
} from 'lambda-layer';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export const deleteSession = async (sessionId: string) => {
  const deleteCommand = new DeleteCommand({
    TableName: sessionsTableOptions.tableName,
    Key: { [sessionsTableOptions.partitionKey.name]: sessionId },
    ReturnValues: 'ALL_OLD',
  });

  const { Attributes } = await ddbDocClient.send(deleteCommand);
  const sessionsItem = Attributes as SessionsTableItem | undefined;

  if (!sessionsItem) {
    throw new Error(`could not find session with id "${sessionId}"`);
  }

  return sessionsItem.refreshToken;
};
