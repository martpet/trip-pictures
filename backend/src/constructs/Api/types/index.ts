export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
