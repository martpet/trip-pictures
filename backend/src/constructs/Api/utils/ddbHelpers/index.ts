export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
