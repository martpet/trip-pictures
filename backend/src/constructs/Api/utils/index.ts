export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
