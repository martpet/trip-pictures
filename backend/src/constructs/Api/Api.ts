import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
