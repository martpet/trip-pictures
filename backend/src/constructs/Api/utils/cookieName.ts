import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
