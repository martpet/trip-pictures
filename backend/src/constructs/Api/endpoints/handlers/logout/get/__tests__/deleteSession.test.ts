import { DeleteCommand, DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';

import { itRejects, itResolves, itSendsAwsCommand } from '~/constructs/Api/utils';

import { deleteSession } from '../deleteSession';

const ddbMock = mockClient(DynamoDBDocumentClient);

const args = ['dummySessionId'] as Parameters<typeof deleteSession>;

beforeEach(() => {
  ddbMock.reset();

  ddbMock
    .on(DeleteCommand)
    .resolves({ Attributes: { refreshToken: 'dummyRefreshToken' } });
});

describe('deleteSession', () => {
  itSendsAwsCommand(DeleteCommand, ddbMock, deleteSession, args);
  itResolves(deleteSession, args);

  describe('when "Attributes" prop is missing from "DeleteCommand" output', () => {
    beforeEach(() => {
      ddbMock.on(DeleteCommand).resolves({});
    });
    itRejects(deleteSession, args);
  });
});
