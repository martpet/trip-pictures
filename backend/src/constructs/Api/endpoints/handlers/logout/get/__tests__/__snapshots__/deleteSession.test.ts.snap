// Vitest Snapshot v1

exports[`deleteSession > resolves with a correct value 1`] = `"dummyRefreshToken"`;

exports[`deleteSession > sends "DeleteCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "Key": {
    "id": "dummySessionId",
  },
  "ReturnValues": "ALL_OLD",
  "TableName": "TP-sessions",
}
`;

exports[`deleteSession > when "Attributes" prop is missing from "DeleteCommand" output > rejects with a correct value 1`] = `[Error: could not find session with id "dummySessionId"]`;
